frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+			Self::process_head(remaining_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register oneself for fast-unstake.
+		///
+		/// The dispatch origin of this call must be signed by the controller account, similar to
+		/// `staking::unbond`.
+		///
+		/// The stash associated with the origin must have no ongoing unlocking chunks. If
+		/// successful, this will fully unbond and chill the stash. Then, it will enqueue the stash
+		/// to be checked in further blocks.
+		///
+		/// If by the time this is called, the stash is actually eligible for fast-unstake, then
+		/// they are guaranteed to remain eligible, because the call will chill them as well.
+		///
+		/// If the check works, the entire staking data is removed, i.e. the stash is fully
+		/// unstaked, and they potentially join a pool with their entire bonded stake.
+		///
+		/// If the check fails, the stash remains chilled and waiting for being unbonded as in with
+		/// the normal staking system, but they lose part of their unbonding chunks due to consuming
+		/// the chain's resources.
+		#[pallet::weight(<T as Config>::WeightInfo::register_fast_unstake())]
+		pub fn register_fast_unstake(
+			origin: OriginFor<T>,
+			maybe_pool_id: Option<PoolId>,
+		) -> DispatchResult {
+			let ctrl = ensure_signed(origin)?;
+
+			let ledger =
+				pallet_staking::Ledger::<T>::get(&ctrl).ok_or(Error::<T>::NotController)?;
+			ensure!(!Queue::<T>::contains_key(&ledger.stash), Error::<T>::AlreadyQueued);
+			ensure!(
+				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
+				Error::<T>::AlreadyHead
+			);
+			// second part of the && is defensive.
+			ensure!(
+				ledger.active == ledger.total && ledger.unlocking.is_empty(),
+				Error::<T>::NotFullyBonded
+			);
+
+			// chill and fully unstake.
+			Staking::<T>::chill(RawOrigin::Signed(ctrl.clone()).into())?;
+			Staking::<T>::unbond(RawOrigin::Signed(ctrl).into(), ledger.total)?;
+
+			// enqueue them.
+			Queue::<T>::insert(ledger.stash, maybe_pool_id);
+			Ok(())
+		}
+
+		/// Deregister oneself from the fast-unstake (and possibly joining a pool).
+		///
+		/// This is useful if one is registered, they are still waiting, and they change their mind.
+		///
+		/// Note that the associated stash is still fully unbonded and chilled as a consequence of
+		/// calling `register_fast_unstake`. This should probably be followed by a call to
+		/// `Staking::rebond`.
+		#[pallet::weight(<T as Config>::WeightInfo::deregister())]
+		pub fn deregister(origin: OriginFor<T>) -> DispatchResult {
+			let ctrl = ensure_signed(origin)?;
+			let stash = pallet_staking::Ledger::<T>::get(&ctrl)
+				.map(|l| l.stash)
+				.ok_or(Error::<T>::NotController)?;
+			ensure!(Queue::<T>::contains_key(&stash), Error::<T>::NotQueued);
+			ensure!(
+				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != stash),
+				Error::<T>::AlreadyHead
```

Comment:
Is it possible for an account id to be not in the queue but be the Head?

Reply:
Yes, and in that time they cannot deregister anymore.